File: src/os/macos/sys.f
\ sys.f — the darwin-arm64 OS seam: syscall numbers + the SVC emitter.
\ Engine emitters say
\ `NR-WRITE SYS,`; porting to another OS/arch swaps this file.

$1002 constant MAP-ANON-PRIVATE
$1012 constant MAP-ANON-PRIVATE-FIXED

$1   constant NR-EXIT
$2   constant NR-FORK
$3   constant NR-READ
$4   constant NR-WRITE
$21  constant NR-ACCESS
$A   constant NR-UNLINK
$F   constant NR-CHMOD
$36  constant NR-IOCTL
$B8  constant NR-SIGRETURN
$5   constant NR-OPEN
$6   constant NR-CLOSE
$25  constant NR-KILL
$52  constant NR-SETPGID
$2A  constant NR-PIPE
$2E  constant NR-SIGACTION
$4A  constant NR-MPROTECT
$5A  constant NR-DUP2
$5C  constant NR-FCNTL
$53  constant NR-SETITIMER
$74  constant NR-GETTIMEOFDAY
$C5  constant NR-MMAP
$E6  constant NR-POLL
$F4  constant NR-SPAWN     \ posix_spawn(&pid, path, 0, 0, argv, envp)
$80  constant NR-RENAME
$88  constant NR-MKDIR
$89  constant NR-RMDIR
$152 constant NR-STAT64
$154 constant NR-LSTAT64
$158 constant NR-GETDIRENTRIES64
$1D9 constant NR-READLINKAT
$1DA constant NR-SYMLINKAT
$7   constant NR-WAIT4     \ wait4(pid, &status, 0, 0)
0    constant NR-EXECVE
0    constant NR-CHDIR
$1   constant NR-EXIT-GROUP

-2 constant AT-FDCWD
0 constant AT-SYMLINK-NOFOLLOW

: SYS, ( n -- )  16 swap MOVZ,  $80 SVC, ;

\ Runtime-emit syscall stencils (TFAM 10 slice 3): the MATCH bad-tag die is
\ EMITTED into the user word, so its write+exit syscall words are baked here at
\ engine-build time rather than assembled by SYS, at build time. The macOS
\ syscall number goes in x16 (movz x16,#NR = $D2800010 + NR*32; Rd and the imm
\ field do not overlap so + == or) and the call is `svc #0x80`. The die never
\ inspects the write return, so the Darwin carry-flag reconciliation SYS, adds is
\ not needed.
$D2800010 NR-WRITE 32 * + constant SYS-EMIT-WRITE       \ movz x16, #NR-WRITE
$D2800010 NR-EXIT-GROUP 32 * + constant SYS-EMIT-EXIT   \ movz x16, #NR-EXIT-GROUP
$D4001001 constant SYS-EMIT-SVC                          \ svc #0x80

: OS-OPEN-RD ( n -- )
   0 swap 0 ADDI,  1 0 MOVZ,  2 0 MOVZ,  NR-OPEN SYS, ;

: OS-OPEN-FLAGS ( -- ) ;

: OS-MMAP-FLAGS ( -- ) ;

File: bootstrap/cg/sys.fs
\ sys.fs -- OS seam selected by HABU_TARGET. Engine emitters say
\ `NR-WRITE SYS,`; target files keep that call surface stable.
\ See docs/porting.md for the full target contract.

require icode.fs

s" HABU_TARGET" getenv
2dup s" linux-aarch64" compare 0= constant HB-TARGET-LINUX?
s" macos-aarch64" compare 0= constant HB-TARGET-MACOS?

HB-TARGET-LINUX? HB-TARGET-MACOS? or 0= [IF]
.( unsupported HABU_TARGET ) cr bye
[THEN]

HB-TARGET-LINUX? [IF]
96 constant IMAGE-TEXT-SIZE-OFF
$1000 constant IMAGE-TEXT-CONTENT-ADJ
0 constant IMAGE-TEXT-TRAILER-ADJ
$22 constant MAP-ANON-PRIVATE
$32 constant MAP-ANON-PRIVATE-FIXED

93  constant NR-EXIT
94  constant NR-EXIT-GROUP
63  constant NR-READ
64  constant NR-WRITE
29  constant NR-IOCTL
139 constant NR-SIGRETURN
56  constant NR-OPEN
57  constant NR-CLOSE
134 constant NR-SIGACTION
226 constant NR-MPROTECT
103 constant NR-SETITIMER
222 constant NR-MMAP
220 constant NR-SPAWN
260 constant NR-WAIT4

: SYS, ( n -- )
   8 swap MOVZ,  0 SVC,
   16 -4095 LIT64,  0 16 CMP, ;

: OS-OPEN-RD ( n -- )
   {: pathreg :}
   1 pathreg 0 ADDI,  0 99 MOVN,  2 0 MOVZ,  3 0 MOVZ,  NR-OPEN SYS, ;

: OS-OPEN-FLAGS ( -- )
   7 3 MOVZ,  6 1 7 AND,
   LBL {: noappend :}
   7 8 MOVZ,  7 1 7 AND,  7 noappend CBZ,
      7 $400 MOVZ,  6 6 7 ORR,
   noappend LBL,
   LBL {: nocreat :}
   7 $200 MOVZ,  7 1 7 AND,  7 nocreat CBZ,
      7 $40 MOVZ,  6 6 7 ORR,
   nocreat LBL,
   LBL {: notrunc :}
   7 $400 MOVZ,  7 1 7 AND,  7 notrunc CBZ,
      7 $200 MOVZ,  6 6 7 ORR,
   notrunc LBL,
   2 6 0 ADDI, ;

: OS-MMAP-FLAGS ( -- )
   7 $12 MOVZ,  6 3 7 AND,
   LBL {: noanon :}
   7 $1000 MOVZ,  7 3 7 AND,  7 noanon CBZ,
      7 $20 MOVZ,  6 6 7 ORR,
   noanon LBL,
   3 6 0 ADDI, ;

[ELSE]
216 constant IMAGE-TEXT-SIZE-OFF
0 constant IMAGE-TEXT-CONTENT-ADJ
$1000 constant IMAGE-TEXT-TRAILER-ADJ
$1002 constant MAP-ANON-PRIVATE
$1012 constant MAP-ANON-PRIVATE-FIXED

1   constant NR-EXIT
1   constant NR-EXIT-GROUP
3   constant NR-READ
4   constant NR-WRITE
54  constant NR-IOCTL
184 constant NR-SIGRETURN
5   constant NR-OPEN
6   constant NR-CLOSE
46  constant NR-SIGACTION
74  constant NR-MPROTECT
83  constant NR-SETITIMER
184 constant NR-SIGRETURN
197 constant NR-MMAP
244 constant NR-SPAWN     \ posix_spawn(&pid, path, 0, 0, argv, envp)
7   constant NR-WAIT4     \ wait4(pid, &status, 0, 0)

: SYS, ( n -- )  16 swap MOVZ,  $80 SVC, ;

: OS-OPEN-RD ( n -- )
   {: pathreg :}
   0 pathreg 0 ADDI,  1 0 MOVZ,  2 0 MOVZ,  NR-OPEN SYS, ;

: OS-OPEN-FLAGS ( -- ) ;

: OS-MMAP-FLAGS ( -- ) ;

[THEN]
